samza-api/src/main/java/org/apache/samza/operators/MessageStream.java

```diff
@@ -16,173 +16,115 @@
-  public <WK, WV, WS extends WindowState<WV>, WM extends WindowOutput<WK, WV>> MessageStream<WM> window(Windows.Window<M, WK, WV, WM> window) {
-    Operator<WM> wndOp = Operators.getWindowOperator(Windows.getInternalWindowFn(window));
-    this.subscribers.add(wndOp);
-    return wndOp.getOutputStream();
-  }
-
-  /**
-   * Method to add an input {@link MessageStream} to a join function. Note that we currently only support 2-way joins.
-   *
-   * @param other  the other stream to be joined w/
-   * @param merger  the common function to merge messages from this {@link MessageStream} and {@code other}
+   * @param otherStream  the other {@link MessageStream} to be joined with
+   * @param joinFn  the function to join {@link Message}s from this and the other {@link MessageStream}
    * @param <K>  the type of join key
-   * @param <JM>  the type of message in the {@link Message} from the other join stream
-   * @param <RM>  the type of message in the {@link Message} from the join function
-   * @return the output {@link MessageStream} from the join function {@code joiner}
+   * @param <OM>  the type of {@link Message}s in the other stream
+   * @param <RM>  the type of {@link Message}s resulting from the join
+   * @return  the joined {@link MessageStream}
    */
-  public <K, JM extends Message<K, ?>, RM extends Message> MessageStream<RM> join(MessageStream<JM> other,
-      BiFunction<M, JM, RM> merger) {
-    MessageStream<RM> outputStream = new MessageStream<>();
-
-    BiFunction<M, JM, RM> parJoin1 = merger::apply;
-    BiFunction<JM, M, RM> parJoin2 = (m, t1) -> merger.apply(t1, m);
-
-    // TODO: need to add default store functions for the two partial join functions
-
-    other.subscribers.add(Operators.<JM, K, M, RM>getPartialJoinOperator(parJoin2, outputStream));
-    this.subscribers.add(Operators.<M, K, JM, RM>getPartialJoinOperator(parJoin1, outputStream));
-    return outputStream;
-  }
+  <K, OM extends Message<K, ?>, RM extends Message> MessageStream<RM> join(MessageStream<OM> otherStream,
+      JoinFunction<M, OM, RM> joinFn);
 
   /**
-   * Method to merge all {@code others} streams w/ this {@link MessageStream}. The merging streams must have the same type {@code M}
+   * Merge all {@code otherStreams} with this {@link MessageStream}.
+   * <p>
+   * The merging streams must have the same {@link Message} type {@code M}.
    *
-   * @param others  other streams to be merged w/ this one
-   * @return  the merged output stream
+   * @param otherStreams  other {@link MessageStream}s to be merged with this {@link MessageStream}
+   * @return  the merged {@link MessageStream}
    */
-  public MessageStream<M> merge(Collection<MessageStream<M>> others) {
-    MessageStream<M> outputStream = new MessageStream<>();
-
-    others.add(this);
-    others.forEach(other -> other.subscribers.add(Operators.getMergeOperator(outputStream)));
-    return outputStream;
-  }
-
+  MessageStream<M> merge(Collection<MessageStream<M>> otherStreams);
```

Comment:
No. I was simply explaining it since Jagadish had a question on this method before.